File: src/core/passes/tail-branch-inversion.ts
import { getOpcode } from "../opcodes.ts";
import type { IrCondition, IrOp } from "../types.ts";
import { hasRewriteBarrier, type IrPass, type IrPassFn } from "./helpers.ts";

const INVERTED_CONDITIONS: Record<IrCondition, { condition: IrCondition; opcode: number }> = {
  "==0": { condition: "!=0", opcode: 0x57 },
  "!=0": { condition: "==0", opcode: 0x5e },
  "<0": { condition: ">=0", opcode: 0x59 },
  ">=0": { condition: "<0", opcode: 0x5c },
};

const run: IrPassFn = (ops, context) => {
  if (context.options.tailBranchInversion !== true) {
    return { ops: [...ops], applied: 0, optimizations: [] };
  }

  const refs = countLabelRefs(ops);
  const result: IrOp[] = [];
  let applied = 0;

  for (let index = 0; index < ops.length; index += 1) {
    const op = ops[index]!;
    if (op.kind !== "cjump" || typeof op.target !== "string" || hasRewriteBarrier(op)) {
      result.push(op);
      continue;
    }

    const jumpIndex = nextExecutableIndex(ops, index + 1);
    if (jumpIndex === undefined) {
      result.push(op);
      continue;
    }
    const jump = ops[jumpIndex]!;
    if (jump.kind !== "jump" || hasRewriteBarrier(jump)) {
      result.push(op);
      continue;
    }

    const labelIndex = jumpIndex + 1;
    const label = ops[labelIndex];
    if (label?.kind !== "label" || label.name !== op.target || (refs.get(label.name) ?? 0) !== 1) {
      result.push(op);
      continue;
    }

    const inverted = INVERTED_CONDITIONS[op.condition];
    result.push({
      ...op,
      condition: inverted.condition,
      target: jump.target,
      opcode: inverted.opcode,
      meta: {
        ...op.meta,
        mnemonic: getOpcode(inverted.opcode).name,
        comment: op.meta.comment?.replace(/^false branch/u, "direct tail branch") ?? "direct tail branch",
      },
      targetMeta: { ...jump.targetMeta },
    });
    index = labelIndex;
    applied += 1;
  }

  if (applied === 0) return { ops: result, applied: 0, optimizations: [] };
  return {
    ops: result,
    applied,
    optimizations: [{
      name: "tail-branch-inversion",
      detail: `Inverted ${applied} branch${applied === 1 ? "" : "es"} whose then-path was only a tail jump.`,
    }],
  };
};

function countLabelRefs(ops: readonly IrOp[]): Map<string, number> {
  const refs = new Map<string, number>();
  for (const op of ops) {
    if (
      (op.kind === "jump" || op.kind === "cjump" || op.kind === "call" || op.kind === "loop") &&
      typeof op.target === "string"
    ) {
      refs.set(op.target, (refs.get(op.target) ?? 0) + 1);
    }
  }
  return refs;
}

function nextExecutableIndex(ops: readonly IrOp[], start: number): number | undefined {
  for (let index = start; index < ops.length; index += 1) {
    if (ops[index]!.kind !== "label") return index;
  }
  return undefined;
}

export const tailBranchInversion: IrPass = {
  name: "tail-branch-inversion",
  run,
  layoutSafe: false,
};

File: native/src/core/passes/tail_branch_inversion.cpp
#include "mkpro/core/passes/tail_branch_inversion.hpp"

#include "mkpro/core/opcodes.hpp"

#include <map>
#include <optional>
#include <string>
#include <utility>
#include <variant>
#include <vector>

namespace mkpro::core::passes {

namespace {

struct InvertedCondition {
  std::string condition;
  int opcode = 0;
};

std::optional<InvertedCondition> inverted_condition(const std::string& condition) {
  if (condition == "==0")
    return InvertedCondition{.condition = "!=0", .opcode = 0x57};
  if (condition == "!=0")
    return InvertedCondition{.condition = "==0", .opcode = 0x5e};
  if (condition == "<0")
    return InvertedCondition{.condition = ">=0", .opcode = 0x59};
  if (condition == ">=0")
    return InvertedCondition{.condition = "<0", .opcode = 0x5c};
  return std::nullopt;
}

std::map<std::string, int> count_label_refs(const std::vector<IrOp>& ops) {
  std::map<std::string, int> refs;
  for (const IrOp& op : ops) {
    if (op.kind != IrKind::Jump && op.kind != IrKind::CondJump && op.kind != IrKind::Call &&
        op.kind != IrKind::Loop) {
      continue;
    }
    if (const auto* target = std::get_if<std::string>(&op.target))
      refs[*target] += 1;
  }
  return refs;
}

} // namespace

PassResult tail_branch_inversion(const std::vector<IrOp>& ops, const PassContext& context) {
  if (!context.options.tail_branch_inversion)
    return PassResult{.ops = ops, .applied = 0, .optimizations = {}};

  const std::map<std::string, int> refs = count_label_refs(ops);
  std::vector<IrOp> result;
  result.reserve(ops.size());
  int applied = 0;

  for (int index = 0; index < static_cast<int>(ops.size()); ++index) {
    const IrOp& op = ops.at(static_cast<std::size_t>(index));
    const auto* target = std::get_if<std::string>(&op.target);
    if (op.kind != IrKind::CondJump || target == nullptr || has_rewrite_barrier(op)) {
      result.push_back(op);
      continue;
    }

    const std::optional<int> jump_index = next_executable_index(ops, index + 1);
    if (!jump_index.has_value()) {
      result.push_back(op);
      continue;
    }
    const IrOp& jump = ops.at(static_cast<std::size_t>(*jump_index));
    if (jump.kind != IrKind::Jump || has_rewrite_barrier(jump)) {
      result.push_back(op);
      continue;
    }

    const int label_index = *jump_index + 1;
    if (label_index >= static_cast<int>(ops.size())) {
      result.push_back(op);
      continue;
    }
    const IrOp& label = ops.at(static_cast<std::size_t>(label_index));
    const auto ref_count = refs.find(label.name);
    if (label.kind != IrKind::Label || label.name != *target || ref_count == refs.end() ||
        ref_count->second != 1) {
      result.push_back(op);
      continue;
    }

    const std::optional<InvertedCondition> inverted = inverted_condition(op.condition);
    if (!inverted.has_value()) {
      result.push_back(op);
      continue;
    }

    IrOp inverted_op = op;
    inverted_op.condition = inverted->condition;
    inverted_op.target = jump.target;
    inverted_op.opcode = inverted->opcode;
    inverted_op.meta.mnemonic = opcode_by_code(inverted->opcode).name;
    if (op.meta.comment == "case mismatch") {
      inverted_op.meta.comment = op.meta.comment;
    } else {
      inverted_op.meta.comment =
          op.meta.comment.has_value() && op.meta.comment->starts_with("false branch")
              ? std::optional<std::string>{std::string("direct tail branch") +
                                           op.meta.comment->substr(std::string("false branch").size())}
              : std::optional<std::string>{"direct tail branch"};
    }
    inverted_op.target_meta = jump.target_meta;
    result.push_back(std::move(inverted_op));
    index = label_index;
    ++applied;
  }

  if (applied == 0)
    return PassResult{.ops = std::move(result), .applied = 0, .optimizations = {}};

  return PassResult{
      .ops = std::move(result),
      .applied = applied,
      .optimizations =
          {
              AppliedOptimization{
                  .name = "tail-branch-inversion",
                  .detail = "Inverted " + std::to_string(applied) + " branch" +
                            (applied == 1 ? "" : "es") +
                            " whose then-path was only a tail jump.",
              },
          },
  };
}

IrPass tail_branch_inversion_pass() {
  return IrPass{
      .name = "tail-branch-inversion",
      .run = tail_branch_inversion,
      .layout_safe = false,
  };
}

} // namespace mkpro::core::passes
